homeassistant/helpers/script.py

```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
     """Throw if script needs to suspend."""
 
 
+class _ScriptRunBase(ABC):
+    """Common data & methods for managing Script sequence run."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        script: "Script",
+        variables: Optional[Sequence],
+        context: Optional[Context],
+    ) -> None:
+        self._hass = hass
+        self._script = script
+        self._variables = variables
+        self._context = context
+        self._step = -1
+        self._action: Optional[Dict[str, Any]] = None
+
+    def _changed(self):
+        self._script._changed()  # pylint: disable=protected-access
+
+    @property
+    def _config_cache(self):
+        return self._script._config_cache  # pylint: disable=protected-access
+
+    @abstractmethod
+    async def async_run(self) -> None:
+        """Run script."""
+
+    async def _async_step(self):
+        try:
+            await getattr(self, f"_async_{_determine_action(self._action)}_step")()
+        except Exception as err:
+            if not isinstance(err, (_SuspendScript, _StopScript)):
```

Comment:
We shouldn't log the exception when it's a BlockingScriptRun either, as the exception will propagate to the caller.

Reply:
Maybe a new method, `should_propagate_exception` should be added. A legacy script run would want to not log if it is seeing an exception prior to a delay, because it will propagate up.

Reply:
I actually think there are two different issues here. The first is who should log the exception, and the second is whether or not the exception should be propagated.

As to the first, I still think it's cleaner to have the script helper log the error rather than requiring the "user" to call back into it (for the same reasons I gave earlier.)

However, I think you have a point on the second issue. It seems clear that if the "user" is still waiting on the script helper then the exception should be propagated up. It also seems clear that if the "user" is _not_ still waiting that the exception should _not_ be propagated.

I'll work on addressing the second point. Let me know if you still agree on the first.

Reply:
It turned out to be pretty easy & clean to control when exceptions should be propagated. 😃

Reply:
I think we should think in "who owns the exception handling". If we're in a blocking call, we should propagate. If we are in a standalone task, the script run should handle it. It is totally fine for automation (or any other future caller) to call back to script to log it, because at least we then give automations the option not to.

Reply:
>I think we should think in "who owns the exception handling".

Regarding the "second" item (i.e., propagating or not), yes, I totally agree.

But regarding the "first" item (who should log it), conceptually, yes, but practically, no.

> It is totally fine for automation (or any other future caller) to call back to script to log it, because at least we then give automations the option not to.

First, the problem, even with the previous implementation, is that if any context switches happen in between (which, granted, doesn't happen now, but who knows in the future), then the "remembered exception" (inside the script helper) could get overwritten by another exception (from another run from another task.) And now that we're allowing multiple, completely independent runs, and that information is in the individual `_XxxScriptRun` object, but the caller only has access to the singular `Script` object, how will `Script.please_log_this_for_me` know _which_ run context (i.e., step # & action) to log the exception against?

Second, the script helper (specifically the `_XxxScriptRun` object) already logs several messages on "its own authority." Why should this be different? It knows how to properly log it, has the full context, and can do so without an otherwise unnecessary extra API.

I suppose one way to address both of these concerns is to decide based on whether or not the caller passes in a logger object. If so, log as before, including logging an exception. If not, don't log anything. But I'm not crazy with this idea because the other script helper log messages are useful.

So I suppose another way is to add a parameter to `Script.async_run` (that it passes down to the `_XxxScriptRun` object) that controls whether or not an exception should be logged. If so then do as it does now. If not then it becomes the caller's _full_ responsibility (i.e., there's no "hey, script helper, can you log this for me" method available, for the reasons I stated above.)

Personally, I still think the way it is (given addressing the propagation of the exception) is the best. The caller doesn't have to worry about whether or not, or how, to log the exception, because it will be done on its behalf. But if you still feel strongly otherwise, I'd go for adding a new parameter to `Script.async_run` and making it the caller's responsibility (and, I suppose, making the script helper _still_ log it if it doesn't propagate the exception.)

Reply:
I looked at how to do this. It's not too bad. I'd add the new param to `Script.__init__` (with a default of `True`) because I think it'd be only one way for each type of caller. That would get passed down to `_XxxScriptRun.__init__` and ultimately to `_ScriptRunBase.__init__`. Then on each call to `_ScriptRunBase._async_step`, if an exception occurred, it would log it if 1) the user asked for it (via the new param), or 2) the exception is not getting propagated (which `_XxxScriptRun.async_run` would tell it.)

I'll go ahead and do this. If the user doesn't specify the new param it would work as it does now.